connectors/sources/onedrive.py

```diff
@@ -0,0 +1,223 @@
+from connectors.source import BaseDataSource
+from connectors.utils import (
+    CacheWithTimeout,
+    CancellableSleeps,
+    RetryStrategy,
+    iso_utc,
+    retryable,
+)
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+BASE_URL = "https://graph.microsoft.com/v1.0/"
+ENDPOINTS = {"PING": "drives"}
+
+
+class TokenRetrievalError(Exception):
+    """Exception class to notify that fetching of access token was not successful."""
+
+    pass
+
+
+class AccessToken:
+    """Class for handling access token for Microsoft Graph APIs"""
+
+    def __init__(self, configuration):
+        self.tenant_id = configuration["tenant_id"]
+        self.client_id = configuration["client_id"]
+        self.client_secret = configuration["client_secret"]
+        self._token_cache = CacheWithTimeout()
+
+    async def get(self):
+        """Get bearer token required for API call.
+
+        If the token is not present in the cache or expired,
+        it calls _set_access_token that sends a POST request
+        to Microsoft for generating a new access token.
+
+        Returns:
+            str: Access Token
+        """
+        if cached_value := self._token_cache.get():
+            return cached_value
+        try:
+            await self._set_access_token()
+        except ClientResponseError as e:
+            match e.status:
+                case 400:
+                    raise TokenRetrievalError(
+                        "Failed to fetch access token. Please verify that provided Tenant ID, Client ID are correct."
+                    ) from e
+                case 401:
+                    raise TokenRetrievalError(
+                        "Failed to fetch access token. Please check if Client Secret is valid."
+                    ) from e
+                case _:
+                    raise TokenRetrievalError(
+                        f"Failed to fetch access token. Response Status: {e.status}, Message: {e.message}"
+                    ) from e
+        return self.access_token
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def _set_access_token(self):
+        """Generate access token with configuration fields and stores it in the cache"""
+        url = f"https://login.microsoftonline.com/{self.tenant_id}/oauth2/v2.0/token"
+        data = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "grant_type": "client_credentials",
+            "scope": "https://graph.microsoft.com/.default",
+        }
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        async with aiohttp.request(
+            method="POST", url=url, data=data, headers=headers, raise_for_status=True
+        ) as response:
+            token_reponse = await response.json()
+            self.access_token = token_reponse["access_token"]
+            self.token_expires_at = datetime.utcnow() + timedelta(
```

Comment:
nit: I feel these are not necessarily instance variable, but local variable.

After `await self._set_access_token()`, you can `return self._token_cache.get()`